### PR TITLE
Check that base branch isn't master in external PR

### PR DESCRIPTION
**Pull request by guyfreund**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19342

## Description
Check that base branch isn't master in external PR

## Does it break backward compatibility?
   - No

## Must have
- [x] Tests
- [ ] Code Review

### CODE REVIEW

**Comment by guyfreund** [comment]
Circle CI won't be green until https://github.com/demisto/dockerfiles/pull/790 will be merged.

File: .circleci/config.yml [hunk 1]
```diff
@@ -107,6 +107,18 @@ jobs:
           name: Run Unit Testing and Lint
           when: always
           command: SKIP_GIT_COMPARE_FILTER=${NIGHTLY} ./Tests/scripts/run_all_pkg_dev_tasks.sh
+      - run:
+          name: Check Base Branch for External PR
```
Comment (Itay4) [inline .circleci/config.yml]:
```suggestion
          name: Verify Base Branch for Contribution
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -107,6 +107,18 @@ jobs:
...
+          when: always
+          command: |
+            if [[ $CIRCLE_BRANCH =~ pull/[0-9]+ ]] ;
+              then
+                python3 ./Tests/scripts/check_base_branch_for_external_pr.py $CIRCLE_BRANCH
+            
+              else
+                echo "Skipping base branch test for internal PRs"
```
Comment (Itay4) [inline .circleci/config.yml]:
```suggestion
                echo "Skipping base branch check for internal PRs"
```

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+def main():
+    circle_branch = sys.argv[1]
+    # If we run this script the circle branch of format "pull/[0-9]+"
+    pr_num = circle_branch.split('pull/')[1]
+    check_base_branch(pr_num)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: no need for `main` function, just do it here

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        sys.exit(1)
+    else:
+        print_color('Base branch of PR num {} is not master - Great!'.format(pr_num), LOG_COLORS.GREEN)
+        sys.exit(0)
+
+
+def main():
+    circle_branch = sys.argv[1]
+    # If we run this script the circle branch of format "pull/[0-9]+"
+    pr_num = circle_branch.split('pull/')[1]
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: use regex instead

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    except requests.exceptions.ConnectionError:
+        # If we didn't succeed to fetch the pr maybe it doesn't exist - then we don't want the build to fail
+        print_warning('Unable to fetch PR num {}'.format(pr_num))
+        return ''
+
+    return ''
+
+
+def check_base_branch(pr_num):
+    print_color('Starting to fetch the base branch of PR num {}'.format(pr_num), LOG_COLORS.GREEN)
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]:
```suggestion
    print_color(f'Fetching the base branch of pull request #{pr_num}', LOG_COLORS.NATIVE)
```

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    base_branch = get_base_branch(pr_num)
+    print_color('Finished to fetch the base branch of PR num {}'.format(pr_num), LOG_COLORS.GREEN)
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: lets remove this, we dont need it

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    if base_branch == 'master':
+        print_error("You cannot merge into master when creating an external PR.")
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]:
```suggestion
        print_error('Cannot merge a contribution directly to master, the pull request reviewer will handle that soon.')
```

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import requests
+import sys
+from Tests.test_utils import print_error, print_warning, print_color, LOG_COLORS
+
+
+def get_base_branch(pr_num):
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: add type hint

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    Fetches the base branch name of PR num {pr_num}
+    :param pr_num: The PR number
+    :return: The name of the base branch if exists
+    """
+
+    # Disable insecure warnings
+    requests.packages.urllib3.disable_warnings()  # pylint: disable=no-member
+
+    try:
+        url = 'https://api.github.com/repos/demisto/content/pulls/{}'.format(pr_num)
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: use f string

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        res = requests.get(url, verify=False)
+        if res.status_code != 200:
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]:
```suggestion
        if not res.ok:
```

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            # If we didn't succeed to fetch the pr maybe it doesn't exist - then we don't want the build to fail
+            print_warning('Unable to fetch PR num {}'.format(pr_num))
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: use f strings and return the api response as well

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            return ''
+        response = res.json()
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: wrap in try-except

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        # Seems like GitHub API return a dict if it's only one PR, but it doesn't mentioned in their API
+        if response and isinstance(response, dict):
+            base = response.get('base', {})
+            base_branch = base.get('ref')
+            if base_branch:
+                return base_branch
+        # GitHub usually returns a list of PRs
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: why would we get a list of PRs ?

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        res = requests.get(url, verify=False)
+        if res.status_code != 200:
+            # If we didn't succeed to fetch the pr maybe it doesn't exist - then we don't want the build to fail
+            print_warning('Unable to fetch PR num {}'.format(pr_num))
+            return ''
+        response = res.json()
+        # Seems like GitHub API return a dict if it's only one PR, but it doesn't mentioned in their API
+        if response and isinstance(response, dict):
+            base = response.get('base', {})
+            base_branch = base.get('ref')
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]:
```suggestion
            return base.get('ref', '')
```

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        # GitHub usually returns a list of PRs
+        elif response and isinstance(response, list) and len(response) == 1:
+            pr = response[0]
+            base = pr.get('base', {})
+            base_branch = base.get('ref')
+            if base_branch:
+                return base_branch
+    except requests.exceptions.ConnectionError:
+        # If we didn't succeed to fetch the pr maybe it doesn't exist - then we don't want the build to fail
+        print_warning('Unable to fetch PR num {}'.format(pr_num))
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]: same here

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+def check_base_branch(pr_num):
+    print_color('Starting to fetch the base branch of PR num {}'.format(pr_num), LOG_COLORS.GREEN)
+    base_branch = get_base_branch(pr_num)
+    print_color('Finished to fetch the base branch of PR num {}'.format(pr_num), LOG_COLORS.GREEN)
+    if base_branch == 'master':
+        print_error("You cannot merge into master when creating an external PR.")
+        sys.exit(1)
+    else:
+        print_color('Base branch of PR num {} is not master - Great!'.format(pr_num), LOG_COLORS.GREEN)
```
Comment (Itay4) [inline Tests/scripts/check_base_branch_for_external_pr.py]:
```suggestion
        print_color('Verified pull request base branch successfully', LOG_COLORS.GREEN)
```

File: Tests/scripts/infrastructure_tests/check_base_branch_for_external_pr_test.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def test_check_base_branch():
+    with requests_mock.Mocker() as m:
+        from Tests.scripts.check_base_branch_for_external_pr import check_base_branch
+        url = 'https://api.github.com/repos/demisto/content/pulls/528'
+        m.get(url, json=bad_response)
+        with pytest.raises(SystemExit) as se:
+            check_base_branch(528)
+            assert se.value.code == 1
```
Comment (Itay4) [inline Tests/scripts/infrastructure_tests/check_base_branch_for_external_pr_test.py]: also assert the error printed

File: Tests/scripts/infrastructure_tests/check_base_branch_for_external_pr_test.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        m.get(url, json=good_response)
+        with pytest.raises(SystemExit) as se:
+            check_base_branch(528)
+            assert se.value.code == 0
```
Comment (Itay4) [inline Tests/scripts/infrastructure_tests/check_base_branch_for_external_pr_test.py]: also assert the success message printed

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 6 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: Do you mean to remove the user msgs?

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 8 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: added type hint in python 2 style.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 5 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: all script must be in python 2.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 9 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: all script must be in python 2.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 11 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: all script must be in python 2.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 13 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: https://developer.github.com/v3/pulls/

look at the response showed here.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 14 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: all script must be in python 2.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 15 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: all script must be in python 2.

File: Tests/scripts/check_base_branch_for_external_pr.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
+import requests
+import sys
+from Tests.test_utils import print_error, print_warning, print_color, LOG_COLORS
+
+
+def get_base_branch(pr_num):
+    """
+    Fetches the base branch name of PR num {pr_num}
```
Comment (ronykoz) [inline Tests/scripts/check_base_branch_for_external_pr.py]: please update to google style docstring

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 11 above)
Comment (ronykoz) [reply Tests/scripts/check_base_branch_for_external_pr.py]: Why does it have to be python2? you are running it with py3..

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 8 above)
Comment (ronykoz) [inline Tests/scripts/check_base_branch_for_external_pr.py]: this function seems long, and it has code duplication, you can add two more functions to make it a lot more clear and shorter code.

File: Tests/scripts/check_base_branch_for_external_pr.py (hunk 11 above)
Comment (guyfreund) [reply Tests/scripts/check_base_branch_for_external_pr.py]: @ronykoz code isn't updated, changed to py3

**Comment by guyfreund** [comment]
For all reviewers:

1. The script is now written in python2, I was trying all day long to edit the script of the stage in circle and make all other tests run with python2 and the test of this pr with python3.
for some reason when I was running in my local env the pytest was running with python3, but in circle it was running with python3.
My last script before the current script was:
```
python -m pytest ./Tests/scripts/hook_validations/tests/ -v
find Tests/scripts/infrastructure_tests -name "*.py" | grep -v check_base_branch | xargs -I{} python -m pytest {} -v
find Tests/scripts/infrastructure_tests -name "*.py" | grep check_base_branch | xargs -I{} python3 -m pytest {} -v
python -m pytest ./Tests/scripts/test_configure_tests.py -v
```

2. In the unit tests I'm trying to test both the exit code and the msg printed to stdout.
The function called inside the test is exiting and then the test is over, in order to test things after the function has finished I'm using pytest.raises(SystemExit), It will also help me to check for the exit code.
My main reference is: [Testing sys.exit() with pytest](https://medium.com/python-pandemonium/testing-sys-exit-with-pytest-10c6e5f7726f), but I've seen a lot of issues handled the same way in other websites.
For some reason, although using pytest.raises the test is finished after calling the function being tested. 

3. All CR comments have been fixed.

**Pull request merged**